### PR TITLE
Fix issues in OSL plugin

### PR DESCRIPTION
**Pull request by usakhelo**
Hi,
These are changes that fix some of the hidden issues OSL plugin:
1. Add Version information to the param block
2. Skip parameters with obsolete flag
3. Add separate variable to keep param ids for maps
4. Start bump parameters IDs with 0 (because they are in separate param block) This may potentially break current scenes but it worked on my tests.
5. Skip parameters with no widget and no connections (Don't remember where this came from but it sounds reasonable. I probably had a case where it failed to add the parameter)
6. Fix getting correct colors from OSL color default values.

Thanks for review!

### CODE REVIEW

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp [hunk 1]
```diff
@@ -426,8 +428,9 @@ void OSLShaderRegistry::create_class_descriptors()
             L"oslTextureMapParams",                     // internal parameter block's name
             0,                                          // ID of the localized name string
             class_descr,                                // class descriptor
-            P_AUTO_CONSTRUCT,                           // block flags
-
+            P_AUTO_CONSTRUCT | P_VERSION,               // block flags
+                                                        // --- P_VERSION arguments ---
+            1,
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: That's supposed to be on the previous line no?

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp [hunk 2]
```diff
@@ -439,12 +442,12 @@ void OSLShaderRegistry::create_class_descriptors()
         for (auto& param_info : shader.m_params)
         {
             param_info.m_max_param.m_max_ctrl_id = ctrl_id++;
-            param_info.m_max_param.m_max_param_id = param_id;
-
-            shader.m_string_map.insert(std::make_pair(string_id, utf8_to_wide(param_info.m_max_param.m_max_label_str)));
 
             if (param_info.m_max_param.m_has_constant)
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: Unrelated to this PR but I wonder: what does "has constant" mean? Should it be `m_is_constant`? (In which case it should also be `m_is_connectable`, etc.)

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp [hunk 3]
```diff
@@ -546,8 +553,8 @@ void OSLShaderRegistry::add_const_parameter(
         if (osl_param.m_has_default)
         {
             const float r = static_cast<float>(osl_param.m_default_value.at(0));
-            const float g = static_cast<float>(osl_param.m_default_value.at(0));
-            const float b = static_cast<float>(osl_param.m_default_value.at(0));
+            const float g = static_cast<float>(osl_param.m_default_value.at(1));
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: That looks like a bug fix? In which case it deserves its own commit.

File: src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.h [hunk 4]
```diff
@@ -86,6 +86,7 @@ class MaxParam
     bool            m_has_constant;
     std::string     m_max_label_str;
     int             m_max_param_id;
+    int             m_max_map_param_id;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.h]: What does the word "map" mean in this context?

Comment (usakhelo) [reply src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.h]: Map is a texture input. Before id of a texture input was calculated based on float or color parameter id, now it's separate variable.

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp (hunk 3 above)
Comment (usakhelo) [reply src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: It is bug fix yes. I'll do separate commit for this.

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp (hunk 2 above)
Comment (usakhelo) [reply src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: Constant is color or float parameter, which can or cannot be textured. I can rename it to m_is_constant

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp (hunk 1 above)
Comment (usakhelo) [reply src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: Probably, since it's only one argument...

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp (hunk 3 above)
Comment (usakhelo) [reply src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: @dictoon just the commit or separate PR for this?

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp (hunk 2 above)
Comment (usakhelo) [reply src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: Now I think `m_is_constant` is misleading because parameter can be a constant and be connectable.. Is it ok you think to have both `m_is_constant` and `m_is_connectable` set to true?

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp (hunk 3 above)
Comment (dictoon) [reply src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: A separate commit is enough!

File: src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.h (hunk 4 above)
Comment (dictoon) [reply src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.h]: Oh ok, makes sense, thanks.

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp (hunk 2 above)
Comment (dictoon) [reply src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: I was just wondering about the choice of verb ("has" vs "is"). I don't know the code enough to know which verb is more accurate.

File: src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.cpp [hunk 5]
```diff
@@ -249,7 +249,7 @@ OSLShaderInfo::OSLShaderInfo(
             max_param.m_param_type = MaxParam::Unsupported;
             max_param.m_page_name = osl_param.m_page;
 
-            max_param.m_has_constant = osl_param.m_valid_default &&
+            max_param.m_is_constant = osl_param.m_valid_default &&
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.cpp]: Formatting suggestion:

    max_param.m_is_constant =
        osl_param.m_valid_default &&
        osl_param.m_lock_geom &&
        osl_param.m_widget != "null";

File: src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.h [hunk 6]
```diff
@@ -83,7 +83,7 @@ class MaxParam
 
     ParamType       m_param_type;
     bool            m_connectable;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.h]: This should be renamed to `m_is_connectable` (if it makes sense).

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp [hunk 7]
```diff
@@ -821,7 +820,7 @@ void OSLShaderRegistry::add_input_parameter(
         auto param_str = utf8_to_wide(max_param.m_osl_param_name + "_map");
         texture_map.push_back(std::make_pair(param_id, utf8_to_wide(max_param.m_max_label_str)));
 
-        const bool short_button = max_param.m_has_constant &&
+        const bool short_button = max_param.m_is_constant &&
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: Formatting suggestion:

    const bool short_button =
        max_param.m_is_constant &&
        (max_param.m_param_type == MaxParam::VectorParam ||
         max_param.m_param_type == MaxParam::NormalParam ||
         max_param.m_param_type == MaxParam::PointParam ||
         max_param.m_param_type == MaxParam::StringPopup ||
         max_param.m_param_type == MaxParam::IntMapper);

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp [hunk 8]
```diff
@@ -460,6 +462,9 @@ void OSLShaderRegistry::create_class_descriptors()
 
             if (param_info.m_max_param.m_connectable)
```
Comment (Mango-3) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: Franz mentioned this before, better to rename it to `m_is_connectable`

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp [hunk 9]
```diff
@@ -486,7 +491,7 @@ void OSLShaderRegistry::create_class_descriptors()
         max_output_param.m_max_ctrl_id = ctrl_id++;
         max_output_param.m_max_param_id = param_id;
         max_output_param.m_connectable = false;
```
Comment (Mango-3) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: Same as above

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp [hunk 10]
```diff
@@ -62,6 +62,7 @@
 
 // Standard headers.
 #include <memory>
+#include <map>
```
Comment (Mango-3) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: Order alphabetically
